torchaudio_contrib/functional.py

```diff
@@ -1,106 +1,133 @@
 
-def _get_freq_values(n_fft, sr):
-    """
-    Get the frequency axis values given the number of FFT bins
-    and sample rate.
-    """
-    return torch.linspace(0, sr/2, n_fft//2 + 1)
-
+    return n_fft, hop_length, window
 
-def get_spectrogram_axis(sig_length, sr, n_fft=2048, hop=512):
-    t = _get_time_values(sig_length, sr, hop)
-    f = _get_freq_values(n_fft, sr)
-    return t, f
 
+def _stft(x, n_fft, hop_length, window, pad, pad_mode, **kwargs):
+    """
+    Wrap torch.stft allowing for multi-channel stft.
 
-def STFT(sig, n_fft=2048, hop=None, window=None, **kwargs):
-    """A wrapper for torch.stft with some preset parameters.
-    Returned value keeps both real and imageinary parts.
-    For STFT magnitude, see spectrogram
+    Args:
+        x (Tensor): Tensor of audio of size (channel, signal)
+            or (batch, channel, signal).
+        n_fft (int): FFT window size.
+        hop_length (int): Number audio of frames between STFT columns.
+        len_win (int): Size of stft window.
+        window (Tensor): 1-D tensor.
+        pad (int): Amount of padding to apply to signal.
+        pad_mode: padding method (see torch.nn.functional.pad).
+        **kwargs: Other torch.stft parameters, see torch.stft for more details.
 
-    input -> (batch, channel, time) or (channel, time)
-    output -> (batch, channel, freq, hop, complex) or (channel, freq, hop, complex)
+    Returns:
+        Tensor: (batch, channel, freq, hop, complex)
+            or (channel, freq, hop, complex)
 
     """
-    if hop is None:
-        hop = n_fft // 4
-    if window is None:
-        window = torch.hann_window(n_fft)
-
-    if sig.dim() == 3:
-        batch, channel, time = sig.size()
-        out_shape = [batch, channel, n_fft//2+1, -1, 2]
-    elif sig.dim() == 2:
-        channel, time = sig.size()
-        out_shape = [channel, n_fft//2+1, -1, 2]
+
+    # (!) Only 3D, 4D, 5D padding with non-constant
+    # padding are supported for now.
+    if pad > 0:
+        x = F.pad(x, (pad, pad), pad_mode)
+
+    if x.dim() == 3:
+        batch, channel, time = x.size()
+        out_shape = [batch, channel, n_fft // 2 + 1, -1, 2]
+    elif x.dim() == 2:
+        channel, time = x.size()
+        out_shape = [channel, n_fft // 2 + 1, -1, 2]
     else:
         raise ValueError('Input tensor dim() must be either 2 or 3.')
 
-    sig = sig.reshape(-1, time)
-    stft = torch.stft(sig, n_fft, hop, window=window, **kwargs)
-    stft = stft.reshape(out_shape)
+    x = x.reshape(-1, time)
+
+    stft_out = torch.stft(x, n_fft, hop_length, window=window,
+                          win_length=window.size(0), **kwargs)
+    stft_out = stft_out.reshape(out_shape)
+
+    return stft_out
+
 
-    return stft
+def stft(x, n_fft=2048, hop_length=None, len_win=None,
+         window=None, pad=0, pad_mode="reflect", **kwargs):
+    """
+    Wraps _stft setting default values and correct window device.
```

Comment:
I don't think the function should magically move the window. `torch.functional.conv2d` will also fail if the input and weights are not on the same device.

Reply:
I think this links back to the outcome of the discussion from [first comment](https://github.com/keunwoochoi/torchaudio-contrib/pull/34/files#r271553604).

Reply:
It's an independent question. If `stft` gets a `window` as its argument, it's currently moved to the device of `x` in the function body. The functional interface should not modify its arguments, the caller is responsible for placing all arguments on the same device, like in `torch.nn.functional`.

Reply:
I meant that if `stft` was in charge of performing the same operations as the layer `STFT` then it also had to be in charge of creating it's own window in case none was passed to it. In that case it would have to be in charge of moving it to the correct device. 

But if we go with the `torch.nn` approach of having functionals do the operations then yes absolutely should not be moving anything to device.